### ISSUE TITLE
HTTP/1.0 performance issues

### ISSUE THREAD

**Issue by jvgutierrez** [envoyproxy/envoy#19821]
*Title*: *HTTP/1.0 performance issues*

*Description*:
At Wikimedia Foundation we are currently evaluating Envoy as a TLS termination layer for our CDN. We've sucessfully deployed it on a 12.5% of our CDN servers but we are running into some issues with HTTP/1.0 clients (PHP 7.3's file_get_contents() use HTTP/1.0 by default). 
*Repro steps*:
As a client I'm using the php:7.3-cli docker image from https://hub.docker.com/_/php running the following code:
```php
<?php
$favicon_md5 = '0860c1e63e95143ca82e59a1cf0cd23b';
$favicon = file_get_contents('https://upload.wikimedia.org/favicon.ico');
if (md5($favicon) === $favicon_md5) {
    echo "favicon fetched successfully\n";
}
```
On the server side we are currently testing envoy 1.18.3. If I target one of the nodes running envoy I get the following response:
```$ time docker run --network host --add-host upload.wikimedia.org:127.0.0.1 -ti php-test
favicon fetched successfully

real    0m20.457s
user    0m0.024s
sys     0m0.024s
```

If I target one of the servers using other TLS termination layer I get:
```
vgutierrez@carrot:~/Dockerfiles/php$ time docker run --network host --add-host upload.wikimedia.org:127.0.0.1 -ti php-test
favicon fetched successfully

real    0m0.433s
user    0m0.022s
sys     0m0.011s
```

Note the difference reported in real time: **20 seconds** VS 0.043 seconds

/clusters output: https://phabricator.wikimedia.org/P20166
/config_dump output: https://phabricator.wikimedia.org/P20167
/server_info output: https://phabricator.wikimedia.org/P20168
/stats output: https://phabricator.wikimedia.org/P20169

**Comment by jvgutierrez** [comment]
Upon further investigation the issue doesn't seem to be related with HTTP/1.0 but with PHP file_get_contents() underlying HTTP client implementation as it can be reproduced as well with HTTP/1.1 by using the following code snippet:
```php
<?php

$favicon_md5 = '0860c1e63e95143ca82e59a1cf0cd23b';
$opts = array('http' =>
    array(
        'protocol_version' => 1.1,
    )
);
$context = stream_context_create($opts);
$favicon = file_get_contents('https://upload.wikimedia.org/favicon.ico', false, $context);
if (md5($favicon) === $favicon_md5) {
    echo "favicon fetched successfully\n";
}
```

and It's finishing after 20s cause that's the current value for delayed_close_timeout on envoy

**Comment by snowp** [comment]
@alyssawilk Do you know what might be going on here?

**Comment by alyssawilk** [comment]
I was about to say it's probably delay-close related.

So HTTP in general can frame the response with one of three ways: content length, chunked encoding, or frame-by-connection-close.

If you don't haven an explicit content length, HTTP/1.1 will chunk, but HTTP/1.0 can only frame by connection close.

Meanwhile, there's another problem which is that if a client is sending data, and the request has not been completely read, a proxy responds with an error and closes the connection, many clients will get a TCP RST (due to uploading after FIN) and not actually read the response.  That race is avoided with delay-close.

It sounds like Envoy could do better detecting if a request is complete, and if so, framing with immediate close and I can pick that up.  In the meantime if there's any way to have your backend set a content length that should work around the problem, or you can lower delay close in the interim.

**Comment by jvgutierrez** [comment]
Thanks for your reply @alyssawilk. I've just retested with HTTP/1.1 and https://upload.wikimedia.org/wikipedia/commons/thumb/9/90/Song_Taizu.jpg/534px-Song_Taizu.jpg as CL is set for that one and we're still seeing the 20 seconds delay

**Comment by alyssawilk** [comment]
You see a 20s delay until close, or a delay until your client processes the response? The delay-close will happen either way, but if the client knows it's waiting on a 40 bytes response it shouldn't have to wait for the FIN before processing the body (that's only a problem if it's framed by connection: close)

**Comment by alyssawilk** [comment]
yeah if your client ignores content length, there's not much you can do other than not delay-close.  You can turn that down or off in your config to deal with your client ignoring HTTP framing :-)

The fix I proposed wouldn't help you because I was only planning on avoiding delay-close if there was no content length. IMO ignoring content length entirely is a client side bug.
